### PR TITLE
feat: add Test Coverage Analyzer recipe (#5119)

### LINKED ISSUE #5119: 🍳 Intermediate Recipe Submission #5
# 🍳 Intermediate Recipe Submission 

Thanks for contributing to the Goose Cookbook!  
This issue represents an **Intermediate Recipe Slot (6–7 points)**.  

Intermediate recipes use **2 extensions working together**. They may also include **1 sub-recipe** for specialization.  

---

## ✅ Acceptance Criteria  
Your recipe will be accepted if it:  
- Uses **2 extensions** (e.g., GitHub + Summarizer, Browser + Developer).  
- Demonstrates integration or chaining between the extensions.  
- Optionally includes **1 sub-recipe** for a specialized task.  
- Includes required YAML fields: `title`, `description`, `instructions`, and at least one `activity`.  
- Validates against the Recipe Reference Guide.  
- Has been tested locally in Goose.  
- Follows the Cookbook Contributing Guide.  

---

## 📋 Submission Checklist  
Before opening your PR, make sure you've:  
- [ ] Chosen a unique and descriptive filename (e.g., `meeting-summarizer.yaml`).  
- [ ] Matched the recipe `title` to the filename.  
- [ ] Confirmed your YAML validates.  
- [ ] Listed both extensions in your recipe.  
- [ ] Included clear instructions and at least one activity.  
- [ ] Verified the recipe works locally in Goose.  

---

## ✨ Inspiration Examples  
*(These are just ideas — you can submit **any real workflow** you've used with Goose.)*  
- GitHub PR summarizer (GitHub + Summarizer).  
- Meeting transcript summarizer (Filesystem + Summarizer).  
- Data cleaner (Filesystem + Developer).  
- API caller + formatter (Web Browser + Developer).  

---

## 📚 Resources  
- [Recipe Reference Guide](https://block.github.io/goose/docs/guides/recipes/recipe-reference)  
- [Cookbook Contributing Guide](https://github.com/block/goose/blob/main/CONTRIBUTING_RECIPES.md#-contributing-recipes-to-goose-cookbook)  
- [Goose Recipes Tutorial (YouTube)](https://youtu.be/8rTliYrQ6Iw)  
- [Existing Recipes](https://github.com/block/goose/tree/main/documentation/src/pages/recipes/data/recipes)  

---

## 🚀 Next Steps  
1. Claim this slot by commenting `.take` below.  
2. Fork the repository and create a new branch for your recipe.
3. Create your recipe file in:  
   `documentation/src/pages/recipes/data/recipes/`  
4. Create a pull request from your fork and reference this issue in the PR description.  

🎉 Once merged, your recipe will be added to the Goose Cookbook!  

---

## 💡 Tips for Success  
- Base your recipe on a real workflow you've tried with Goose.  
- Show how the two extensions interact — don't just use them separately.  
- Keep the workflow useful and reproducible.

### PR DESCRIPTION
**Pull request by ARYPROGRAMMER**
## Summary

Intermediate recipe that analyzes test coverage gaps and generates actionable recommendations on where to add tests for maximum impact. Identifies critical untested code paths, high-complexity functions with low coverage, and provides specific test case suggestions with examples.

**Key Features:**

- **2 Extensions Used**: Developer (test execution, code analysis), Memory (baseline tracking)
- **4 Customizable Parameters**: target_path, coverage_tool, min_coverage, exclude_patterns
- **Multi-Framework Support**: Auto-detects Jest, Pytest, Go test, Cargo, and more
- **Smart Prioritization**: Ranks testing needs by criticality, complexity, and recent changes
- **4 Priority Levels**: Critical, High, Medium, Low - based on coverage, complexity, and business impact
- **Actionable Output**: Specific test case recommendations with examples and edge cases
- **Lightweight Memory**: Stores only summary metrics (coverage %, date, top 3 files) for progress tracking
- **9-Step Workflow**: Memory Check → Detect Framework → Run Coverage → Analyze Gaps → Prioritize → Generate Suggestions → Create Report → Store Metrics → Present Findings
- **Markdown Report**: Comprehensive `test-coverage-report.md` with coverage tables, gap analysis, and improvement tracking

## Type of Change

- [x] Feature
- [ ] Bug fix
- [ ] Refactor / Code quality
- [ ] Performance improvement
- [ ] Documentation
- [ ] Tests
- [ ] Security fix
- [ ] Build / Release
- [ ] Other

## Testing

**Primary test command:**
```bash
goose run --recipe documentation/src/pages/recipes/data/recipes/test-coverage-analyzer.yaml --params target_path=./test-coverage-sample --params coverage_tool=jest --params min_coverage=80
```

**Additional Test command:**
```bash
# Analyze specific directory with custom threshold
goose run --recipe documentation/src/pages/recipes/data/recipes/test-coverage-analyzer.yaml --params target_path=./crates/goose --params min_coverage=70 --params coverage_tool=auto

# Analyze entire project with custom exclusions
goose run --recipe documentation/src/pages/recipes/data/recipes/test-coverage-analyzer.yaml --params target_path=. --params exclude_patterns=test,spec,mock,dist,build
```

## Screenshots
demo (2)
demo (3)
demo (1)

Manual Testing Completed:
✅ Validated YAML structure and syntax (passed)
✅ Confirmed all required fields present (version, title, description, instructions, prompt, activities)
✅ Verified 2 extensions properly configured (developer, memory)
✅ Confirmed parameter substitution in prompt works correctly
✅ Validated Jinja2 templating with min_coverage and exclude_patterns variables
✅ Verified memory extension stores only lightweight summary data (3 values: coverage %, date, critical files)
✅ Tested multi-framework detection with Jest
✅ Confirmed prioritization workflow: Critical (0% coverage on utils.js) → High (calculator.js functions)
✅ Verified markdown report generation
✅ Tested with sample project achieving 9.75% coverage and identifying gaps correctly

## Related Issues
Fixes #5119 

Email: **notaryasingh@gmail.com**

### CODE REVIEW

**Comment by ARYPROGRAMMER** [comment]
can someone review this as well as my other prs. seems pretty non complicated and simple. thanks

File: documentation/src/pages/recipes/data/recipes/test-coverage-analyzer.yaml [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+     ## High Priority Gaps
+     ...
+     
+     ## Improvement Tracking
+     - Previous Coverage: [Retrieved from memory if available]
+     - Current Coverage: X%
+     - Change: +/- Y%
+     ```
+  
+  8. **Store Summary Metrics in Memory**
```
Comment (iandouglas) [inline documentation/src/pages/recipes/data/recipes/test-coverage-analyzer.yaml]: I don't see a purpose to storing this in a memory, per my other PR review for daily standups, it would be better if you directed goose to read the previous summary of notes than to keep these stats in a growing list of goose memories that will clutter up other LLM interactions with unnecessary context.

**Comment by taniacryptid** [comment]
> can someone review this as well as my other prs. seems pretty non complicated and simple. thanks

Thank you for your patience in waiting for review. The team was away for the extended weekend in the states and are back reviewing today!

Wanted to let you know the team is reviewing your PRs today. However, please note my [comment here](https://github.com/block/goose/pull/5125#issuecomment-3402443383) that you've gone against the rules and self-assigned yourself 5 recipe submission issues at once, when the rules do state within each issue to only assign yourself one at a time, like you mentioned to [another community member here](https://github.com/block/goose/issues/5078#issuecomment-3383290203).

**Comment by ARYPROGRAMMER** [comment]
> > can someone review this as well as my other prs. seems pretty non complicated and simple. thanks
> 
> Thank you for your patience in waiting for review. The team was away for the extended weekend in the states and are back reviewing today!
> 
> Wanted to let you know the team is reviewing your PRs today. However, please note my [comment here](https://github.com/block/goose/pull/5125#issuecomment-3402443383) that you've gone against the rules and self-assigned yourself 5 recipe submission issues at once, when the rules do state within each issue to only assign yourself one at a time, like you mentioned to [another community member here](https://github.com/block/goose/issues/5078#issuecomment-3383290203).

that is one of my concerns and i thought that maybe someone will either unassign me or keep some of my prs at draft. I am willing to close any of my prs as requested by the team. @taniandjerry 

let me know further

**Comment by ARYPROGRAMMER** [comment]
> This recipe is a neat idea, but I feel that the testing suites that devs use are already going to have full HTML-based summaries that are clickable to see areas of code coverage, etc.. While this would give a good summary, I don't feel it's a viable tool for most devs. I think it would be far fewer tokens used to summarize an existing coverage report than to go run tests AND analyze the code as well. Maybe we could change this to tell goose to run the tests with coverage, and summarize that report instead?
> 
> Also, I don't see a good use case here for the memory extension, I think that should be removed.

Thanks for the feedback, @iandouglas, I've completely refactored the recipe to address your concerns:

Changes made:

- Refocused on report summarization - Runs tests with coverage and summarizes the HTML/JSON reports rather than manually analyzing code
-  Removed manual complexity analysis - All analysis comes from the coverage report data
-  Minimal memory usage - Memory now stores ONLY the overall coverage percentage (< 50 characters total: "Project coverage: 82%")
- Points users to HTML reports - Acknowledges test frameworks generate comprehensive HTML coverage reports for detailed line-by-line views

The memory extension now serves a single, lightweight purpose: track simple improvement ("Coverage improved from 75% to 82%") without polluting context with detailed data. I think now this gives more reason for the devs to use it

Is this approach acceptable?

**Comment by taniacryptid** [comment]
Closing some recipe submissions from @ARYPROGRAMMER and accepting their best recipes in single PR here #5125 ! Thank you so much for your contributions.
